client/adventureMap/AdventureMapInterface.cpp

```diff
@@ -649,10 +656,10 @@ void AdventureMapInterface::onTileHovered(const int3 &mapPos)
 			}
 		case SpellID::DIMENSION_DOOR:
 			{
-				const TerrainTile * t = LOCPLINT->cb->getTile(mapPos, false);
-				int3 hpos = LOCPLINT->localState->getCurrentArmy()->getSightCenter();
-				if((!t || t->isClear(LOCPLINT->cb->getTile(hpos))) && isInScreenRange(hpos, mapPos))
-					CCS->curh->set(Cursor::Map::TELEPORT);
+				const TerrainTile * t = LOCPLINT->cb->getTileForDimensionDoor(mapPos, LOCPLINT->localState->getCurrentHero());
+				int3 heroPosition = LOCPLINT->localState->getCurrentArmy()->getSightCenter();
+				if(t && t->isClear(LOCPLINT->cb->getTile(heroPosition))/* && isInScreenRange(hpos, mapPos)*/)
+					CCS->curh->set(Cursor::Map::TELEPORT); //TODO: something wrong with beyond east spell range border cursor on arrogance after TP-ing near underground portal on previous day
```

Comment:
IMO this entire block should be moved to DimensionDoorMechanics / ScuttleBoatMechanics
for example `bool canCastOnTile(int3 mapTile)` method

This will fix two issues:
- apparently, neither Scuttle Boat nor DD check for screen range on server. Yay, cheating.
- with this check, your CGameInfoCallback::getTileForDimensionDoor method will be unnecessary

Reply:
what about the fact that this logic is still needed by clientside anyway for determining cursor type etc.?